libglusterfs/src/gf-io.c

```diff
@@ -0,0 +1,1626 @@
+
+    return next;
+}
+
+static void
+gf_io_push_ready(gf_io_list_item_t *item)
+{
+    CMM_STORE_SHARED(item->next, GF_IO_REQUEST_READY);
+}
+
+/* Get a new pending request to process. If there are no pending requests
+ * available for some amount of time, the pending queue is resetted so
+ * that other threads could manage it when new requests arrive. */
+static gf_io_list_item_t *
+gf_io_push_next(gf_io_worker_t *worker, gf_io_list_item_t *item)
+{
+    gf_io_list_item_t *next, *tmp;
+
+    /* Actively wait for little bit to see if more requests are coming
+     * (or the next one is still being queued) before releasing the
+     * ownership of the pending requests queue. */
+    next = gf_io_push_wait_limited(worker, item);
+    if (caa_unlikely(next == NULL)) {
+        /* After some time there are no more requests. The most likely
+         * cause is that the queue is empty. We try to reset it. */
+        tmp = uatomic_cmpxchg(&gf_io.sq.queue, item, NULL);
+        if (caa_unlikely(tmp != item)) {
+            /* More requests detected. We need to wait until the request
+             * is fully inserted into the queue. */
+            next = gf_io_push_wait(worker, item);
+        }
+    }
+
+    /* Mark the old item as completely processed so that other threads
+     * can process the completion results. */
+    gf_io_push_ready(item);
+
+    return next;
+}
+
+/* Submit one request to the kernel. */
+static void
+gf_io_push_one(gf_io_worker_t *worker, gf_io_request_t *req)
+{
+    int32_t res;
+
+    res = gf_io_sqe_add(req);
+    if (caa_unlikely(res < 0)) {
+        /* In the unlikely case that we have been unable to obtain an SQE
+         * for this request, we call the completion callback right now
+         * passing the error code. */
+        req->res = res;
+
+        gf_io_cbk(worker, req);
+    }
+}
+
+/* Send requests to the kernel. */
+void
+gf_io_push(gf_io_worker_t *worker, gf_io_list_item_t *item)
+{
+    do {
+        gf_io_push_one(worker, gf_io_object(item, gf_io_request_t, list));
+
+        /* Keep processing entries until the queue is empty. */
+    } while ((item = gf_io_push_next(worker, item)) != NULL);
+}
+
+static int32_t
+gf_io_push_delayed(gf_io_request_t *req)
+{
+    if (caa_unlikely(req->worker == NULL)) {
+        /* This can only happen if there has been an error sending the
+         * delayed request from a non worker thread. In that case, we
+         * try to immediately process the pending requests. */
+        gf_io_push(NULL, req->data);
+    } else {
+        /* Otherwise we attach the list of pending requests to the worker
+         * so that they are executed after processing all completions.
+         * This prevents potentially long delays while processing
+         * completions. */
+        req->worker->delayed = req->data;
+    }
+}
+
+/* Send requests to the kernel in the background. */
+void
+gf_io_delayed(gf_io_worker_t *worker, gf_io_list_item_t *item)
+{
+    gf_io.sq.delay.res = 0;
```

Comment:
nit: perhaps we can use a local variable of gf_io_.sq.delay here.

Reply:
Since we are not dereferencing any pointer all these accesses are resolved at compile time, so they are "free" from the performance point of view. Anyway I can do that for more clarity.